src/msac.rs

```diff
@@ -176,27 +182,59 @@ impl Default for MsacContext {
     }
 }
 
-impl MsacContext {
-    fn set_buf(&mut self, buf: &[u8]) {
-        let Range { start, end } = buf.as_ptr_range();
-        self.buf_pos = start;
-        self.buf_end = end;
+impl MsacAsmContext {
+    fn allow_update_cdf(&self) -> bool {
+        self.allow_update_cdf != 0
     }
 
-    fn with_buf(&mut self, mut f: impl FnMut(&[u8]) -> &[u8]) {
-        // # Safety
-        //
-        // [`Self::buf_pos`] and [`Self::buf_end`] are the start and end ptrs of the `buf` slice,
-        // and are only set in [`Self::set_buf`], which derives them from a valid slice.
+    fn with_buf<'a>(&mut self, mut f: impl FnMut(&'a [u8]) -> &'a [u8]) {
+        // SAFETY: [`Self::buf_pos`] and [`Self::buf_end`] are the start and end ptrs of the [`MsacContext::data`] slice.
+        // They are initially set in [`MsacContext::new`].
+        // [`Self::buf_pos`] may be modified in [`MsacContext::with_asm`] calls, but it remains valid.
+        // This [`Self::with_buf`] method itself is always called with [`MsacContext::with_asm`] higher on the call stack,
+        // so the owner of [`Self::buf_pos`], [`MsacContext::data`], is still alive and the lifetime is valid.
         let buf = unsafe {
             let len = self.buf_end.offset_from(self.buf_pos) as usize;
-            std::slice::from_raw_parts(self.buf_pos, len)
+            slice::from_raw_parts(self.buf_pos, len)
         };
-        self.set_buf(f(buf));
+        let Range { start, end } = f(buf).as_ptr_range();
+        self.buf_pos = start;
+        self.buf_end = end;
     }
+}
 
-    fn allow_update_cdf(&self) -> bool {
-        self.allow_update_cdf != 0
+#[derive(Default)]
+pub struct MsacContext {
+    asm: MsacAsmContext,
+    pub data: Option<CArc<[u8]>>,
+    pub data_offset: usize,
+}
+
+impl Deref for MsacContext {
+    type Target = MsacAsmContext;
+
+    fn deref(&self) -> &Self::Target {
+        &self.asm
+    }
+}
+
+impl DerefMut for MsacContext {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.asm
+    }
+}
+
+impl MsacContext {
+    fn with_asm<R>(&mut self, mut f: impl FnMut(&mut MsacAsmContext) -> R) -> R {
+        let data = self.data.as_ref().unwrap().as_ref();
+        // Hold this reference across `f`, which ensures [`MsacAsmContext::with_buf`] is safe.
+        let retval = f(&mut self.asm);
```

Comment:
How does holding the reference here make `with_buf` safe? I would think that the thing that makes it safe is that we're enforcing that `buf_pos` and `buf_end` stay within bounds of the source allocation.

Reply:
Well we also need to make sure the data `buf_pos` can't be freed while we're accessing it.  Holding the reference is what guarantees that.

Reply:
How would `buf_pos` get freed in a way that holding onto `data` across the call to `f` would prevent? `f` is getting the `MsacAsmContext` which doesn't have access to `data`, which is what actually owns the buffer pointed to by `buf_pos` right? So I don't see how the call to `f` would be able to free the buffer other than doing something unsafe like passing `buf_pos` to `free`, which wouldn't even work right?

Reply:
How? Isn't the data owned by the `CArc` which we're still holding onto? It doesn't seem like holding a reference to `self.data` is what's preventing it from being freed, it's the fact that we still own a copy of the `CArc` and hence are preventing the ref count from going to 0.

Reply:
Hmm, yeah that's true.

Reply:
Doing it a different way now.